### PR TITLE
Utilities/StaticAnalysers: Fix segfault in FiniteMathChecker. Fix garbage values in CMS code rules checkers category.

### PR DESCRIPTION
**Pull request by gartung**
Fix garbage values in CMS code rules checkers categories introduced by the incorrect use of std::unique_ptr with function temporaries.

Fix segfault in FiniteMathChecker caused by used os clang::ento::CheckerBase::getCheckerName().

Testing locally.

### CODE REVIEW

**Comment by gartung** [comment]
code-checks

File: Utilities/StaticAnalyzers/src/ArgSizeChecker.cpp [hunk 1]
```diff
@@ -70,11 +70,11 @@ namespace clangcms {
...
               clang::ento::PathDiagnosticLocation::createBegin(PVD, ctx.getSourceManager());
-
-          std::unique_ptr<BugType> BT = std::make_unique<clang::ento::BugType>(
-              this, "Function parameter copied by value with size > max", "ArgSize");
+          if (!BT1)
+            BT1 = std::make_unique<clang::ento::BugType>(
+                this, "Function parameter copied by value with size > max", "ArgSize");
           std::unique_ptr<BasicBugReport> report =
-              std::make_unique<BasicBugReport>(*BT, llvm::StringRef(os.str()), DLoc);
+              std::make_unique<BasicBugReport>(*BT1, llvm::StringRef(os.str()), DLoc);
```
Comment (makortel) [inline Utilities/StaticAnalyzers/src/ArgSizeChecker.cpp:77]: Assuming the `BasicBugReport` does not take address of the first argument, and that `clang::ento::BugType` is not a heavy type, how about something along
```cpp
clang::ento::BugType bt1(this, "Function parameter copied by value with size > max", "ArgSize");
auto report = std::make_unique<BasicBugReport>(std::move(bt1), llvm::StringRef(os.str()), DLoc);
```
or
```suggestion
              std::make_unique<BasicBugReport>(clang::ento::BugType(this, "Function parameter copied by value with size > max", "ArgSize"),
                                               llvm::StringRef(os.str()), DLoc);
```
?

Comment (gartung) [reply Utilities/StaticAnalyzers/src/ArgSizeChecker.cpp:77]: OK

Comment (gartung) [reply Utilities/StaticAnalyzers/src/ArgSizeChecker.cpp]: Memory usage hits the limits on cmsdev25 when I converted all of the BugType constructors to this.

Comment (makortel) [reply Utilities/StaticAnalyzers/src/ArgSizeChecker.cpp:77]: Yeah, after looking at it I see `clang::ento::BugReport` (base class of `BasicBugReport`) constructor taking  the address of the first argument
https://clang.llvm.org/doxygen/BugReporter_8h_source.html
so indeed the earlier you had is needed. Sorry for the noise and extra work.

**Comment by gartung** [comment]
please build

**Comment by gartung** [comment]
please test

File: Utilities/StaticAnalyzers/src/ArgSizeChecker.h [hunk 2]
```diff
@@ -4,13 +4,16 @@
 #include <clang/StaticAnalyzer/Core/Checker.h>
 #include <clang/StaticAnalyzer/Core/PathSensitive/CheckerContext.h>
 #include <clang/StaticAnalyzer/Core/BugReporter/BugType.h>
+#include "FWCore/Utilities/interface/thread_safety_macros.h"
```
Comment (makortel) [inline Utilities/StaticAnalyzers/src/ArgSizeChecker.h:7]: This `#include` would require
```xml
<use name="FWCore/Utilities" source_only="1"/>
```
in `BuildFile.xml`.

Comment (gartung) [reply Utilities/StaticAnalyzers/src/ArgSizeChecker.h:7]: That header is used in other checker headers. Never knew it had to be declared.

**Comment by gartung** [comment]
Still producing garbage valuesfor category name

```


 
--
Àafe function cal0 | 1 |  
pû | 1 |  
setCache Globache | 2 |  
algoBeginJobcalled | 2 |  
compareUsingrnys | 11 |  
connect_front>> | 5 |  
getSuperLayerNamea | 1 |  
@±Óã–afe function cal0 | 1 |  
non-const global static variable | 1588 |  
non-const static variable | 9336 |  

```

**Comment by gartung** [comment]
Running static analysis locally to make sure reverting the change fixes the garbage catrgory names.

**Comment by gartung** [comment]
@makortel The category names are back to what they should be.

**Comment by makortel** [comment]
+core

**Comment by perrotta** [comment]
+1

**Pull request merged**